### DISCUSSION TITLE
Python function內的local與global variable

### DISCUSSION THREAD

**Discussion by gnsJhenJie** [1101-GS4719-C/Python-Community-GS4719-C#572] [Q&A] [answered]
請問以下這段code為什麼會在第4行出錯 UnboundLocalError: local variable 'a' referenced before assignment ，我的想法是在執行第4行時的a應該是指global variable的a，但看錯誤訊息後發現python似乎把第4行的a當成local variable了
```python
a = "hello"

def foo():
    print(a)     # 這行出錯
    a = "123"
    print(a)

foo()
```
<img width="672" alt="Screen Shot 2022-04-27 at 15 59 27" src="https://user-images.githubusercontent.com/32983347/165470392-6f49a614-7350-4eca-96b6-950e0648bb75.png">

**Comment by JSHT** [comment, marked as answer]
感覺有點像 Hoisting 的概念。剛剛找到 [這篇文章](https://ithelp.ithome.com.tw/questions/10202228)，不知道能不能作為解釋。

**Reply by gnsJhenJie** [reply]
看起來應該是這篇文章的概念沒錯。原本以為Python的interpreter是由上到下一行一行執行的，所以上面的code不會被下面影響，不過看來是在define function的時候interpreter就會對整個function內的code做一些識別了。

**Reply by gnsJhenJie** [reply]
這邊有找到一個更詳細的講Python Scope Resolution的文章  [Scope Resolution in Python | LEGB Rule - GeeksforGeeks ](https://www.geeksforgeeks.org/scope-resolution-in-python-legb-rule/) 大家也可以參考一下

**Comment by chia-kai-chang** [comment]
### OK
```python
a = "hello"

def foo():
    print(a)    # global variable
    b = "123"
    print(b)    # local variable

foo()
```


### OK
```python
a = "hello"

def foo():
    print(a)    # global variable

foo()
```


### OK
```python
a = "hello"

def foo():
    global a
    print(a)    # global variable
    a = "123"    # global variable
    print(a)    # global variable

foo()
```


### Error
```python
a = "hello"

def foo()
    print(a)     # 原以為 a 是 global variable
    a = "123"    # a 被指定為 local variable，跟上一行程式牴觸
    print(a)

foo()
```
> 要先決定好，函式裡面的 `a` 是 global variable 還是 local variable，也就是要確定變數 `a` 的 Namespace 為何，只能擇一

**Reply by chia-kai-chang** [reply]
![image](https://user-images.githubusercontent.com/1768891/165552723-c4aca904-47cb-4fc4-b603-b48549cd1491.jpeg)

![image](https://user-images.githubusercontent.com/1768891/165552762-742b994d-307d-45f1-b37f-45b1a9b80bbf.jpeg)

![image](https://user-images.githubusercontent.com/1768891/165552799-2f4093b0-2308-4883-b0b7-ad6529713ad1.jpeg)

Ref: Real World Instrumentation with Python
